### PR TITLE
[+] Provide alternative nearest neighbor backends for dknn

### PR DESCRIPTION
**Pull request by nartes**
I propose to replace falconn with faiss. It speeds up an index construction as well as a search.
There's no mentioning of falconn in Docker file nor in .sh script related.
I'm using the following to install faiss with gpu speed up on Ubuntu container with Conda.
```bash
conda install -c pytorch faiss-gpu
```
I've disabled random samples shuffling upon the index construction in faiss.
But multiple runs for me given different **dknn_preds**.
Is it supposed to be deterministic?
I can't check it with falconn since it's too slow.

### CODE REVIEW

**Comment by npapernot** [comment]
Thank you for the PR. I'm all for extending the code base to include new knn libraries but in order to keep the archival status of the dknn example, could you add a flag that specifies which library to use for nearest neighbor search (e.g., falconn, fais, ...) and then keep the falconn code as the default library? I'll take a look at the code afterwards. Thanks!

**Comment by nartes** [comment]
Yeah, sure.

**Comment by nartes** [comment]
Here is the output of multiple runs of dknn_tutorial with
  lsh_backend: FAISS
  number_bits: 64
```
1.

(9250, 10)                                                                                                                             
0.9742702702702702                                                                                                                                                                                                                                                            
Saving reliability diagram at: /tmp/dknn_test.pdf                                                                                                                                                                                                                             
(9250,) (9250,)                                                                                                                                                                                                                                                               
[0.7891816920943134, 0.9473684210526315, 0.9841269841269841, 0.9921798631476051, 0.992822966507177, 0.9989868287740629, 1.0, 0.9987293519695044, 1.0, 1.0]                                                                                                                    
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[721, 1007, 1071, 1023, 836, 987, 890, 787, 955, 973]                                                                                                                                                                                                                         
9250                                                                                                                                                                                                                                                                          
(9250, 10)                                                                                                                                                                                                                                                                    
0.24010810810810812                                                                                                                                                                                                                                                           
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                                                                                                                                                              
(9250,) (9250,)                                                                                                                                                                                                                                                               
[0.19800498753117207, 0.24965197215777263, 0.2377963737796374, 0.2912621359223301, 0.2860824742268041, 0.41605839416058393, 0.43790849673202614, 0.39361702127659576, 0.36633663366336633, 0.08695652173913043]                                                               
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[4010, 2155, 1434, 618, 388, 274, 153, 94, 101, 23]
9250

2.
(9250, 10)
0.9712432432432433
Saving reliability diagram at: /tmp/dknn_test.pdf
(9250,) (9250,)
[0.7699248120300752, 0.9310344827586207, 0.9861386138613861, 0.988056460369164, 0.9905123339658444, 0.99625, 0.9969666329625885, 1.0, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[665, 1044, 1010, 921, 1054, 800, 989, 814, 1059, 894]
9250
(9250, 10)
0.20724324324324325
Saving reliability diagram at: /tmp/dknn_adv.pdf
(9250,) (9250,)
[0.16778523489932887, 0.1879507848960543, 0.2251759186864738, 0.2509157509157509, 0.2660753880266075, 0.3191489361702128, 0.2916666666666667, 0.42105263157894735, 0.37583892617449666, 0.34782608695652173]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[3725, 2357, 1279, 546, 451, 282, 240, 152, 149, 69]
9250

3.
(9250, 10)
0.9704864864864865
Saving reliability diagram at: /tmp/dknn_test.pdf
(9250,) (9250,)
[0.7744565217391305, 0.9344422700587084, 0.9823651452282157, 0.9885974914481186, 0.9950207468879668, 0.9957310565635006, 0.9987864077669902, 1.0, 0.9978991596638656, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[736, 1022, 964, 877, 1205, 937, 824, 925, 952, 808]
9250
(9250, 10)
0.22086486486486487
Saving reliability diagram at: /tmp/dknn_adv.pdf
(9250,) (9250,)
[0.1705937794533459, 0.20517316966242877, 0.24709042076991944, 0.3304042179261863, 0.2962962962962963, 0.3632286995515695, 0.3732394366197183, 0.48214285714285715, 0.6052631578947368, 0.6296296296296297]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[4244, 2281, 1117, 569, 459, 223, 142, 112, 76, 27]
9250

4.
(9250, 10)
0.9726486486486486
Saving reliability diagram at: /tmp/dknn_test.pdf
(9250,) (9250,)
[0.7360655737704918, 0.9472318339100346, 0.9815005138746146, 0.9922737306843267, 0.9954285714285714, 0.9988385598141696, 1.0, 0.9987966305655837, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[610, 1156, 973, 906, 875, 861, 1018, 831, 957, 1063]
9250
(9250, 10)
0.21275675675675676
Saving reliability diagram at: /tmp/dknn_adv.pdf
(9250,) (9250,)
[0.17803138373751784, 0.19968429360694553, 0.1963470319634703, 0.20314735336194564, 0.27111111111111114, 0.34576271186440677, 0.45535714285714285, 0.49137931034482757, 0.4842105263157895, 0.5]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[3505, 2534, 1314, 699, 450, 295, 224, 116, 95, 18]
9250

5.
(9250, 10)
0.9713513513513513
Saving reliability diagram at: /tmp/dknn_test.pdf
(9250,) (9250,)
[0.7933673469387755, 0.9353876739562624, 0.9821605550049554, 0.9936775553213909, 0.9954389965792474, 0.9903381642512077, 0.9973992197659298, 1.0, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[784, 1006, 1009, 949, 877, 828, 769, 1128, 948, 952]
9250
(9250, 10)
0.2427027027027027
Saving reliability diagram at: /tmp/dknn_adv.pdf
(9250,) (9250,)
[0.21751896256422804, 0.2397935241670577, 0.2147458840372226, 0.26194690265486725, 0.33424657534246577, 0.336, 0.4476744186046512, 0.4166666666666667, 0.4166666666666667, 0.2903225806451613]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[4087, 2131, 1397, 565, 365, 250, 172, 168, 84, 31]
9250

6.
(9250, 10)
0.9695135135135136
Saving reliability diagram at: /tmp/dknn_test.pdf
(9250,) (9250,)
[0.8133472367049009, 0.928150765606596, 0.9755813953488373, 0.9896421845574388, 0.995945945945946, 0.9971428571428571, 0.9947229551451188, 0.9991721854304636, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[959, 849, 860, 1062, 740, 700, 758, 1208, 1166, 948]
9250
(9250, 10)
0.25113513513513513
Saving reliability diagram at: /tmp/dknn_adv.pdf
(9250,) (9250,)
[0.20701592430186938, 0.23297297297297298, 0.2720465890183028, 0.3031161473087819, 0.3413333333333333, 0.3561643835616438, 0.35784313725490197, 0.5049019607843137, 0.49056603773584906, 0.39215686274509803]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[4333, 1850, 1202, 706, 375, 219, 204, 204, 106, 51]
9250

```

File: cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py [hunk 1]
```diff
@@ -16,6 +16,8 @@
 import matplotlib.pyplot as plt
 import numpy as np
 from six.moves import xrange
+import faiss
```
Comment (npapernot) [inline cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py]: Can you make the imports conditional on the corresponding flag value? (so users only have to install the library that they want to use)

File: cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py [hunk 2]
```diff
@@ -413,8 +550,18 @@ def get_activations(data):
 def main(argv=None):
   assert dknn_tutorial()
 
-
-if __name__ == '__main__':
+def define_default_argument_values():
```
Comment (npapernot) [inline cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py]: Could you remove `define_default_argument_values` and leave the flags declarations as they were previously for consistency with the rest of the library?

**Comment by npapernot** [comment]
Regarding the non-deterministic results on FAISS, I have never used the library so I can't really comment on  that

**Comment by nartes** [comment]
> Regarding the non-deterministic results on FAISS, I have never used the library so I can't really comment on that

Could you please provide multiple runs output with falconn?
Just to compare.

**Comment by npapernot** [comment]
The code I've been using is the one you are editing in this example

**Comment by nartes** [comment]
> The code I've been using is the one you are editing in this example

How many cores are needed for falconn to perform reasonably well?
With 2x2.4 GHz it takes 20 minutes or more per run.

**Comment by nartes** [comment]
I think in couple hours I'll provide falconn runs as well here.

**Comment by nartes** [comment]
```
1.
(9250, 10)                                                         
0.9872432432432432                                                                                                                     
Saving reliability diagram at: /tmp/dknn_test.pdf 
(9250,) (9250,)                                                    
[0.7611940298507462, 0.958974358974359, 0.9789915966386554, 0.9914163090128756, 0.9920948616600791, 0.9924433249370277, 0.9973684210526316, 1.0, 0.9954954954954955, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]       
[335, 390, 476, 466, 253, 397, 380, 306, 444, 5803]                
9250                                                               
(9250, 10)                                                         
0.32075675675675674                                                
Saving reliability diagram at: /tmp/dknn_adv.pdf                   
(9250,) (9250,)                                                    
[0.2983072500798467, 0.20353356890459365, 0.2904929577464789, 0.4346153846153846, 0.5346534653465347, 0.5631067961165048, 0.6071428571428571, 0.6, 0.8059701492537313, 0.8664495114006515]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]
[6262, 1415, 568, 260, 101, 103, 112, 55, 67, 307]
9250

2.
(9250, 10)                                                                                                                             
0.9876756756756757                                                 
Saving reliability diagram at: /tmp/dknn_test.pdf                                                                                      
(9250,) (9250,)                                                    
[0.7816091954022989, 0.9612244897959183, 0.9769230769230769, 0.9856115107913669, 0.9948979591836735, 1.0, 1.0, 0.9961240310077519, 1.0, 0.9998292349726776]                                                                                                                   
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[348, 490, 390, 417, 392, 275, 380, 258, 444, 5856]                                                                                                                                                                                                                           
9250                                                               
(9250, 10)                                                         
0.3067027027027027                                                 
Saving reliability diagram at: /tmp/dknn_adv.pdf                   
(9250,) (9250,)                                                                                                                        
[0.3065626434144103, 0.21056149732620322, 0.26695842450765866, 0.39545454545454545, 0.4090909090909091, 0.4375, 0.5, 0.55, 0.6595744680851063, 0.7486338797814208]                                                                                                            
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[6537, 1496, 457, 220, 132, 64, 74, 40, 47, 183]                                                                                                                                                                                                                              
9250

3.
(9250, 10)                                                         
0.9887567567567568                                                 
Saving reliability diagram at: /tmp/dknn_test.pdf                  
(9250,) (9250,)                                                    
[0.8126888217522659, 0.9534883720930233, 0.9686800894854586, 0.9871794871794872, 0.9939577039274925, 0.9971988795518207, 1.0, 1.0, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[331, 430, 447, 390, 331, 357, 403, 293, 405, 5863]                                                                                    
9250                                                                                                                                                                                                                                                                          
(9250, 10)                                                                                                                             
0.33254054054054055                                                                                                                                                                                                                                                           
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                       
(9250,) (9250,)                                                                                                                        
[0.3140144073346431, 0.21556122448979592, 0.30985915492957744, 0.4349775784753363, 0.5135135135135135, 0.6287878787878788, 0.6448598130841121, 0.5769230769230769, 0.8125, 0.8484848484848485]                                                                                
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[6108, 1568, 568, 223, 148, 132, 107, 52, 80, 264]                                                                                     
9250

4.
(9250, 10)                                                                                                                                                                                                                                                                    
0.9878918918918919                                                                                                                     
Saving reliability diagram at: /tmp/dknn_test.pdf                                                                                                                                                                                                                             
(9250,) (9250,)                                                                                                                        
[0.75625, 0.9684684684684685, 0.9767441860465116, 0.9877049180327869, 0.9883720930232558, 0.9972451790633609, 1.0, 1.0, 0.9977064220183486, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[320, 444, 344, 488, 344, 363, 398, 287, 436, 5826]                                                                                    
9250                                                                                                                                   
(9250, 10)                                                                                                                             
0.3516756756756757                                                 
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                       
(9250,) (9250,)                                                                                                                        
[0.35621128138245844, 0.23251643753735804, 0.26153846153846155, 0.2916666666666667, 0.4049586776859504, 0.43820224719101125, 0.6052631578947368, 0.5111111111111111, 0.7407407407407407, 0.8857142857142857]                                                                  
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[6134, 1673, 455, 288, 121, 89, 76, 45, 54, 315]                                                                                       
9250  

5.
(9250, 10)                                                                                                                             
0.988                                                                                                                                  
Saving reliability diagram at: /tmp/dknn_test.pdf                  
(9250,) (9250,)                                                                                                                        
[0.7923976608187134, 0.9465478841870824, 0.9844357976653697, 0.9770773638968482, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]                         
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[342, 449, 514, 349, 366, 258, 383, 294, 462, 5833]                                                                                    
9250                                                                                                                                   
(9250, 10)                                                                                                                             
0.32497297297297295                                                
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                       
(9250,) (9250,)                                                                                                                        
[0.33488971180164223, 0.1894167167769092, 0.2736842105263158, 0.35353535353535354, 0.3951612903225806, 0.5517241379310345, 0.4823529411764706, 0.5909090909090909, 0.5111111111111111, 0.8492063492063492]                                                                    
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[6211, 1663, 570, 198, 124, 58, 85, 44, 45, 252]                                                                                       
9250                                             

6.
(9250, 10)                                                                                                                             
0.988                                                                                                                                  
Saving reliability diagram at: /tmp/dknn_test.pdf                  
(9250,) (9250,)                                                                                                                        
[0.7850746268656716, 0.9553990610328639, 0.9763948497854077, 0.9879518072289156, 0.9963636363636363, 0.9945205479452055, 1.0, 1.0, 1.0, 0.9998297292695386]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[335, 426, 466, 415, 275, 365, 397, 262, 436, 5873]                                                                                                                                                                                                                           
9250                                                                                                                                                                                                                                                                          
(9250, 10)                                                                                                                             
0.3348108108108108                                                 
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                       
(9250,) (9250,)                                                                                                                        
[0.3203498927215712, 0.21857585139318886, 0.3190789473684211, 0.396, 0.46534653465346537, 0.5164835164835165, 0.5632183908045977, 0.6486486486486487, 0.7, 0.875]                                                                                                             
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[6059, 1615, 608, 250, 101, 91, 87, 37, 50, 352]                                                                                       
9250             

7.
(9250, 10)                                                                                                                             
0.9869189189189189                                                                                                                     
Saving reliability diagram at: /tmp/dknn_test.pdf                                                                                      
(9250,) (9250,)                                                                                                                        
[0.7507418397626113, 0.9566265060240964, 0.970703125, 0.9938837920489296, 0.9939393939393939, 1.0, 1.0, 1.0, 1.0, 1.0]                 
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[337, 415, 512, 327, 330, 396, 421, 297, 437, 5778]                                                                                    
9250                                                                                                                                   
(9250, 10)                                                                                                                             
0.2690810810810811                                                 
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                       
(9250,) (9250,)                                                                                                                                                                                                                                                               
[0.26541341405129837, 0.17516902274124155, 0.2545126353790614, 0.2541436464088398, 0.38028169014084506, 0.41904761904761906, 0.5, 0.627906976744186, 0.6122448979591837, 0.7970297029702971]                                                                                  
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                                                                                                                                                                  
[6277, 1627, 554, 181, 142, 105, 70, 43, 49, 202]                                                                                      
9250             
```

**Comment by npapernot** [comment]
Results look very different for the first few bins when comparing FAISS runs to FALCONN runs, are the parameters for both libraries comparable?

**Comment by nartes** [comment]
```
1.                              
(9250, 10)                                              
0.9863783783783784                                      
Saving reliability diagram at: /tmp/dknn_test.pdf                                                                                                                                                                                                                             
(9250,) (9250,)                                                                                                 
[0.7294117647058823, 0.9655870445344129, 0.9726775956284153, 0.9908883826879271, 0.9917127071823204, 1.0, 1.0, 1.0, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[340, 494, 366, 439, 362, 254, 409, 267, 460, 5859]
9250                                                                                                            
(9250, 10)                                              
0.33037837837837836                                     
Saving reliability diagram at: /tmp/dknn_adv.pdf                                                                                       
(9250,) (9250,)                                                                                                 
[0.32244833569963716, 0.19945909398242057, 0.2647058823529412, 0.3697478991596639, 0.4084507042253521, 0.39705882352941174, 0.5161290322580645, 0.4528301886792453, 0.7951807228915663, 0.8587896253602305]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[6339, 1479, 408, 238, 142, 68, 93, 53, 83, 347]                   
9250                                                               

2.                                              
(9250, 10)                                                         
0.9863783783783784                                                 
Saving reliability diagram at: /tmp/dknn_test.pdf                  
(9250,) (9250,)                                                    
[0.7294117647058823, 0.9655870445344129, 0.9726775956284153, 0.9908883826879271, 0.9917127071823204, 1.0, 1.0, 1.0, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[340, 494, 366, 439, 362, 254, 409, 267, 460, 5859]                
9250                                                               
(9250, 10)                                                         
0.33037837837837836                                                
Saving reliability diagram at: /tmp/dknn_adv.pdf                   
(9250,) (9250,)                                                    
[0.32244833569963716, 0.19945909398242057, 0.2647058823529412, 0.3697478991596639, 0.4084507042253521, 0.39705882352941174, 0.5161290322580645, 0.4528301886792453, 0.7951807228915663, 0.8587896253602305]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[6339, 1479, 408, 238, 142, 68, 93, 53, 83, 347]                   
9250                                                               

3.                                                     
(9250, 10)                                                         
0.9863783783783784                                                 
Saving reliability diagram at: /tmp/dknn_test.pdf                  
(9250,) (9250,)                                                    
[0.7294117647058823, 0.9655870445344129, 0.9726775956284153, 0.9908883826879271, 0.9917127071823204, 1.0, 1.0, 1.0, 1.0, 1.0]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[340, 494, 366, 439, 362, 254, 409, 267, 460, 5859]                
9250                                                               
(9250, 10)                                                         
0.33037837837837836                                                
Saving reliability diagram at: /tmp/dknn_adv.pdf                   
(9250,) (9250,)                                                    
[0.32244833569963716, 0.19945909398242057, 0.2647058823529412, 0.3697478991596639, 0.4084507042253521, 0.39705882352941174, 0.5161290322580645, 0.4528301886792453, 0.7951807228915663, 0.8587896253602305]
[0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95]                                                                           
[6339, 1479, 408, 238, 142, 68, 93, 53, 83, 347]                   
9250                                                              
```

**Comment by npapernot** [comment]
OK to ignore CLA, I edited the PR

**Comment by nartes** [comment]
Cool, no more questions left then? Just wait for the travis script?

**Comment by npapernot** [comment]
Yes I am just waiting on Travis, thanks a lot for the PR. The results are sufficiently consistent between FALCONN and FAISS for us to merge the PR. There is not much value in tuning the individual parameters of each approach given that optimal values of these parameters will depend on the model architecture and dataset too.

**Comment by nartes** [comment]
@npapernot Do you know any practical application of dknn for big enough models?

**Comment by npapernot** [comment]
I have only applied to CNNs with <= 10 layers, but I don't think more layers should be a problem as long as the FALCONN/FAISS implementation can handle the nearest-neighbor search in each layer.

**Pull request merged**